# This is the tag along code for the stage B quiz

### Author details
#### Name: Jimoh Abdulsomad Abiola
#### Hamoye ID:  15e47f79f981f000

#### Importing the necessary modules

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.preprocessing import MinMaxScaler,StandardScaler
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split

In [2]:
file = pd.read_csv("energydata_complete.csv")
file.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9  T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53   6.60        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56   6.48        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50   6.37        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40   6.25        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40   6.13        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410390  45.410390  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

### Deriving insights from the data

In [3]:
file.shape

(19735, 29)

In [4]:
file.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.412580    755.522602   
std    ...      2.014712      4.151497      5.318464      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.670000    750.933333   
50%    ...     19.390000     40.900000      6.920000    756.100000   
75%    ...     20.600000     44.338095     10.400000    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760995     24.988033   
std       14.901088      2.451221     11.794719      4.195248     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.430000     24.897653   
75%       91.666667      5.500000     40.000000      6.570000     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

In [5]:
file.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

<b> The dataset has 19,735 rows and 29 columns (from file.shape)<p><p>
<b>The dataset contains 29 columns which is seen to have one "object" type ("date"), two "int64" type ("Appliances" and "lights") and the rest are "float64" datatype (from file.info())

#### Question 12
<b>From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [6]:
x = file[["T2"]]
y = file.T6
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.3, random_state = 42)
model_t = LinearRegression()
model_t.fit(x_train,y_train)


LinearRegression()

In [7]:
pred_t = model_t.predict(x_test)
r2 = r2_score(y_test,pred_t)
round(r2,2)

0.64

#### Question 13
<b>Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. Answer the following questions:

What is the Mean Absolute Error (in two decimal places)?

In [8]:
file = file.drop(columns = ["date","lights"])
file.head()

Appliances     T1       RH_1    T2       RH_2     T3       RH_3         T4  \
0          60  19.89  47.596667  19.2  44.790000  19.79  44.730000  19.000000   
1          60  19.89  46.693333  19.2  44.722500  19.79  44.790000  19.000000   
2          50  19.89  46.300000  19.2  44.626667  19.79  44.933333  18.926667   
3          50  19.89  46.066667  19.2  44.590000  19.79  45.000000  18.890000   
4          60  19.89  46.333333  19.2  44.530000  19.79  45.000000  18.890000   

        RH_4         T5  ...         T9   RH_9  T_out  Press_mm_hg  RH_out  \
0  45.566667  17.166667  ...  17.033333  45.53   6.60        733.5    92.0   
1  45.992500  17.166667  ...  17.066667  45.56   6.48        733.6    92.0   
2  45.890000  17.166667  ...  17.000000  45.50   6.37        733.7    92.0   
3  45.723333  17.166667  ...  17.000000  45.40   6.25        733.8    92.0   
4  45.530000  17.200000  ...  17.000000  45.40   6.13        733.9    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  
2   6.333333   55.333333        5.1  28.642668  28.642668  
3   6.000000   51.500000        5.0  45.410390  45.410390  
4   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 27 columns]

In [9]:
scaler = MinMaxScaler()
scaled = scaler.fit_transform(file)
norm = pd.DataFrame(scaled, columns = file.columns)
norm.head()

Appliances       T1      RH_1        T2      RH_2        T3      RH_3  \
0    0.046729  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066   
1    0.046729  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871   
2    0.037383  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569   
3    0.037383  0.32735  0.524080  0.225345  0.678414  0.215188  0.758685   
4    0.046729  0.32735  0.531419  0.225345  0.676727  0.215188  0.758685   

         T4      RH_4        T5  ...        T9      RH_9     T_out  \
0  0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1  0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369132   
2  0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365595   
3  0.341441  0.770949  0.175506  ...  0.219563  0.671909  0.361736   
4  0.341441  0.762697  0.178691  ...  0.219563  0.671909  0.357878   

   Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0     0.097674  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1     0.100000  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2     0.102326  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  
3     0.104651  0.894737   0.428571    0.776923   0.524887  0.908261  0.908261  
4     0.106977  0.894737   0.404762    0.717949   0.520362  0.201611  0.201611  

[5 rows x 27 columns]

In [10]:
model_m = LinearRegression()
x_n = norm.drop(columns = ["Appliances"])
y_n = norm["Appliances"]
nx_train, nx_test, ny_train, ny_test = train_test_split(x_n,y_n, test_size = 0.3, random_state = 42)

model_m.fit(nx_train,ny_train)
pred_m = model_m.predict(nx_test)

In [11]:
mae = mean_absolute_error(ny_test,pred_m)
round(mae,2)

0.05

#### Question 14
<b>What is the Residual Sum of Squares (in two decimal places)?

In [12]:
rss = np.sum(np.square(ny_test-pred_m))
round(rss,2)

45.34

#### Question 15
<b>What is the Root Mean Squared Error (in three decimal places)?

In [13]:
rmse = np.sqrt(mean_squared_error(ny_test,pred_m))
round(rmse,3)

0.088

#### Question 16
<b>What is the Coefficient of Determination (in two decimal places)?

In [14]:
#r2_score is the coefficient of determination 
r2_m = r2_score(ny_test,pred_m)
round(r2_m,2)

0.15

#### Question 17
<b> Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [15]:
def get_weight_df(model, feat, col_name):
    weights = pd.Series(model.coef_,feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ["Features",col_name]
    weights_df[col_name].round(3)
    return weights_df

In [16]:
model_m_weight = get_weight_df(model_m,nx_train,"Weights")
model_m_weight

Features       Weights
0           rv1 -2.709069e+11
1          RH_2 -4.567765e-01
2         T_out -3.219795e-01
3            T2 -2.363138e-01
4            T9 -1.900100e-01
5          RH_8 -1.576251e-01
6        RH_out -7.769583e-02
7          RH_7 -4.456069e-02
8          RH_9 -3.981388e-02
9            T5 -1.561445e-02
10           T1 -3.241639e-03
11  Press_mm_hg  6.846967e-03
12           T7  1.030132e-02
13   Visibility  1.230447e-02
14         RH_5  1.598570e-02
15         RH_4  2.639800e-02
16           T4  2.910551e-02
17    Windspeed  2.920193e-02
18         RH_6  3.802484e-02
19         RH_3  9.609399e-02
20           T8  1.020178e-01
21    Tdewpoint  1.178497e-01
22           T6  2.364801e-01
23           T3  2.905649e-01
24         RH_1  5.535431e-01
25          rv2  2.709069e+11

RV1 and RV2 are the lowest and highest respctively, but, cannot be found in the options

#### Question 18
<b>Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [17]:
model_r = Ridge(alpha = 0.4)
model_r.fit(nx_train,ny_train)
pred_r = model_r.predict(nx_test)
rmse_r = np.sqrt(mean_squared_error(ny_test,pred_r))
round(rmse_r,2)

0.09

#### Question 19
<b>Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [18]:
model_l = Lasso(alpha = 0.001)
model_l.fit(nx_train,ny_train)
model_l_weight = get_weight_df(model_l,nx_train,"Lasso Weights")
model_l_weight

Features  Lasso Weights
0        RH_out      -0.049557
1          RH_8      -0.000110
2            T1       0.000000
3     Tdewpoint       0.000000
4    Visibility       0.000000
5   Press_mm_hg      -0.000000
6         T_out       0.000000
7          RH_9      -0.000000
8            T9      -0.000000
9            T8       0.000000
10         RH_7      -0.000000
11          rv1      -0.000000
12           T7      -0.000000
13           T6       0.000000
14         RH_5       0.000000
15           T5      -0.000000
16         RH_4       0.000000
17           T4      -0.000000
18         RH_3       0.000000
19           T3       0.000000
20         RH_2      -0.000000
21           T2       0.000000
22         RH_6      -0.000000
23          rv2      -0.000000
24    Windspeed       0.002912
25         RH_1       0.017880

In [19]:
# Can be achieved through boolean masking
non_zero =model_l_weight[model_l_weight["Lasso Weights"] != 0]
non_zero

Features  Lasso Weights
0      RH_out      -0.049557
1        RH_8      -0.000110
24  Windspeed       0.002912
25       RH_1       0.017880

In [20]:
len(non_zero)

4

#### Question 20
<b>What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)

In [21]:
pred_l = model_l.predict(nx_test)
rmse_l = np.sqrt(mean_squared_error(ny_test,pred_r))
round(rmse_l,3)

0.088